### PR TITLE
Better binary<->decimal floating-point conversion

### PR DESCRIPTION
**Pull request by klausler**
This change extracts the code for converting binary floating-point to/from decimal from lib/evaluate/real.* and repackages it into its own library so that it can be used by both the f18 compiler and its runtime for formatted Fortran I/O floating-point editing.

The new `decimal.h` header file has APIs for both C and C++ to enable the code to be used as a general facility.

The algorithms have been corrected, benchmarked, optimized, and tested; new tests have been added, including a thorough sweep through all 32-bit IEEE FP values.

Fortran rounding modes are supported.  A flag that enables minimal decimal digit string formatting was implemented (with the assumption that decimal->binary conversion will use the same binary format and default rounding) for future use with list-directed and namelist output.

### CODE REVIEW

File: test/evaluate/testing.cc [hunk 1]
```diff
@@ -108,8 +107,8 @@ FailureDetailPrinter Compare(const char *file, int line, const char *xs,
     return BitBucket;
   } else {
     ++failures;
-    fprintf(stderr, "%s:%d: FAIL: %s[0x%llx] %s %s[0x%llx]\n", file, line, xs,
-        x, rel, ys, y);
+    fprintf(stderr, "%s:%d: FAIL: %s[0x%jx] %s %s[0x%jx]\n", file, line, xs,
+        static_cast<std::uint64_t>(x), rel, ys, static_cast<std::uint64_t>(y));
```
Comment (tskeith) [inline test/evaluate/testing.cc]: These casts should be to `std::uintmax_t` as above.

File: lib/decimal/decimal.h [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// C and C++ API for binary-to/from-decimal conversion package.
+
+#ifndef FORTRAN_DECIMAL_H_
+#define FORTRAN_DECIMAL_H_
```
Comment (tskeith) [inline lib/decimal/decimal.h]: I think the include guard should be `FORTRAN_DECIMAL_DECIMAL_H_`.

File: test/evaluate/testing.cc (hunk 1 above)
Comment (klausler) [reply test/evaluate/testing.cc]: Thanks!

File: lib/decimal/decimal.h (hunk 2 above)
Comment (klausler) [reply lib/decimal/decimal.h]: Yes, of course.

File: lib/decimal/binary-floating-point.h [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+  BinaryFloatingPointNumber() {}  // zero
+  BinaryFloatingPointNumber(const BinaryFloatingPointNumber &that) = default;
+  BinaryFloatingPointNumber(BinaryFloatingPointNumber &&that) = default;
+  BinaryFloatingPointNumber &operator=(
+      const BinaryFloatingPointNumber &that) = default;
+  BinaryFloatingPointNumber &operator=(
+      BinaryFloatingPointNumber &&that) = default;
+
+  template<typename A> explicit constexpr BinaryFloatingPointNumber(A x) {
+    std::memcpy(reinterpret_cast<void *>(&raw),
```
[the remark this answers is not in the document]
Comment (klausler) [reply lib/decimal/binary-floating-point.h]: Of course; thanks.

File: lib/decimal/binary-floating-point.h [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+#include <cstring>
+#include <type_traits>
+
+namespace Fortran::decimal {
+
+template<int BITS> struct HostUnsignedIntTypeHelper {
+  using type = std::conditional_t<(BITS <= 8), std::uint8_t,
+      std::conditional_t<(BITS <= 16), std::uint16_t,
+          std::conditional_t<(BITS <= 32), std::uint32_t,
+              std::conditional_t<(BITS <= 64), std::uint64_t, __uint128_t>>>>;
```
[the remark this answers is not in the document]
Comment (klausler) [reply lib/decimal/binary-floating-point.h]: We don't support 32-bit modes at all (early design decision).

File: lib/decimal/decimal-to-binary.cc [hunk 5]
```diff
@@ -0,0 +1,431 @@
...
+  case 'Q': {
+    bool negExpo{*++q == '-'};
+    if (*q == '-' || *q == '+') {
+      ++q;
+    }
+    if (*q >= '0' && *q <= '9') {
+      int expo{0};
+      for (int j{0}; j < 8 && *q >= '0' && *q <= '9'; ++j) {
+        expo = 10 * expo + *q++ - '0';
+      }
```
[the remark this answers is not in the document]
Comment (klausler) [reply lib/decimal/decimal-to-binary.cc]: I'll refine this a bit.

File: lib/evaluate/real.cc [hunk 6]
```diff
@@ -446,8 +473,21 @@ std::ostream &Real<W, P, IM>::AsFortran(std::ostream &o, int kind) const {
     if (parenthesize) {
       o << "(";
     }
-    Decimal<Real> decimal;
-    o << decimal.FromReal(*this).ToString() << '_' << kind;
+    using B = decimal::BinaryFloatingPointNumber<P>;
+    const auto *value{reinterpret_cast<const B *>(this)};
+    char buffer[24000];  // accommodate real*16
+    auto result{decimal::ConvertToDecimal<P>(buffer, sizeof buffer,
+        static_cast<decimal::DecimalConversionFlags>(0), 999999,
```
[the remark this answers is not in the document]
Comment (klausler) [reply lib/evaluate/real.cc]: I was being lazy and didn't want to cast the result of the `sizeof`.  Will fix.

File: lib/decimal/decimal.h [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+extern "C" {
+#define NS(x) Fortran::decimal::x
+#else
+#define NS(x) x
+#endif /* C++ */
+
+NS(ConversionToDecimalResult)
+ConvertFloatToDecimal(char *, size_t, enum NS(DecimalConversionFlags),
+    int digits, enum NS(FortranRounding), float);
+NS(ConversionToDecimalResult)
```
[the remark this answers is not in the document]
Comment (klausler) [reply lib/decimal/decimal.h]: Yes.

File: lib/decimal/decimal.h [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+    int digits, enum NS(FortranRounding), double);
+#if __x86_64__
+NS(ConversionToDecimalResult)
+ConvertLongDoubleToDecimal(char *, size_t, enum NS(DecimalConversionFlags),
+    int digits, enum NS(FortranRounding), long double);
+#endif
+
+NS(ConversionResultFlags)
+ConvertDecimalToFloat(const char **, float *, enum NS(FortranRounding));
+NS(ConversionResultFlags)
```
[the remark this answers is not in the document]
Comment (klausler) [reply lib/decimal/decimal.h]: Yes.

File: lib/decimal/decimal.h [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// C and C++ API for binary-to/from-decimal conversion package.
+
+#ifndef FORTRAN_DECIMAL_DECIMAL_H_
+#define FORTRAN_DECIMAL_DECIMAL_H_
+
+#include "binary-floating-point.h"
```
[the remark this answers is not in the document]
Comment (klausler) [reply lib/decimal/decimal.h]: Oh, good point.  Will do.

**Comment by klausler** [comment]
I have addressed all of your comments (thanks for them!), but I'll wait until you have had a chance to review them before merging.

**Comment by jeanPerier** [comment]
Changes look good, thanks for addressing the comments. I could use `decimal.h` in a C program with gcc/clang and the very long exponent example now runs in 0.007s, that's a nice 10000 speed-up !

**Pull request merged**